compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt

```diff
@@ -0,0 +1,503 @@
+                        ),
+                        nameHint = "constructor",
+                        irType = symbols.javaLangReflectConstructor.defaultType
+                    )
+                +constructorSetAccessible(irGet(constructorVar))
+                +constructorNewInstance(irGet(constructorVar), call.getValueArguments())
+            }
+
+    private fun generateReflectiveFieldGet(
+        declaringClass: IrType,
+        fieldName: String,
+        fieldType: IrType,
+        instance: IrExpression?, // null ==> static field on `declaringClass`
+        symbol: IrSymbol,
+    ): IrExpression =
+        context.createJvmIrBuilder(symbol)
+            .irBlock(resultType = fieldType) {
+                val classVar = createTmpVariable(
+                    javaClassObject(declaringClass),
+                    nameHint = "klass",
+                    irType = symbols.kClassJava.owner.getter!!.returnType
+                )
+                val fieldVar = createTmpVariable(
+                    getDeclaredField(irGet(classVar), fieldName),
+                    nameHint = "field",
+                    irType = symbols.javaLangReflectField.defaultType
+                )
+                +fieldSetAccessible(irGet(fieldVar))
+                +fieldGet(irGet(fieldVar), instance ?: irGet(classVar))
+            }
+
+    private fun generateReflectiveFieldGet(getField: IrGetField): IrExpression =
+        generateReflectiveFieldGet(
+            getField.symbol.owner.parentClassOrNull!!.defaultType,
+            getField.symbol.owner.name.asString(),
+            getField.type,
+            getField.receiver,
+            getField.symbol
+        )
+
+    private fun generateReflectiveFieldSet(
+        declaringClass: IrType,
+        fieldName: String,
+        value: IrExpression,
+        type: IrType,
+        instance: IrExpression?,
+        symbol: IrSymbol
+    ): IrExpression {
+        return context.createJvmIrBuilder(symbol)
+            .irBlock(resultType = type) {
+                val fieldVar =
+                    createTmpVariable(
+                        getDeclaredField(
+                            javaClassObject(declaringClass),
+                            fieldName
+                        ),
+                        nameHint = "field",
+                        irType = symbols.javaLangReflectField.defaultType
+                    )
+                +fieldSetAccessible(irGet(fieldVar))
+                +fieldSet(irGet(fieldVar), instance ?: irNull(), value)
+            }
+    }
+
+    private fun generateReflectiveFieldSet(setField: IrSetField): IrExpression =
+        generateReflectiveFieldSet(
+            setField.symbol.owner.parentClassOrNull!!.defaultType,
+            setField.symbol.owner.name.asString(),
+            setField.value,
+            setField.type,
+            setField.receiver,
+            setField.symbol,
+        )
+
+    @OptIn(ObsoleteDescriptorBasedAPI::class)
```

Comment:
@udalov This is just about ready for review, barring this little remaining obstacle :)

The information of whether the accessor is a default accessor is lost in translation between front- and back-end, where PropertyAccessorDescriptors are mapped to IrSimpleFunctions, which do not contain the information.

A possible place to put the information may be the `correspondingPropertySymbol?.owner` of `accessor`, as a property on `IrProperty`.

What do you think?

Reply:
Answered in internal Slack channel